### PR TITLE
ENT-8936: Added support for multiple superhubs per feeder

### PR DESCRIPTION
**Pull request by craigcomstock**
NOT backwards compatible change to save feeder dumps
to a superhub hostkey specific directory and then
have superhubs pull dumps in the usual way from that
alternate directory.

Ticket: ENT-8936
Changelog: title

### CODE REVIEW

File: templates/federated_reporting/dump.sh [hunk 1]
```diff
@@ -76,3 +76,13 @@ else
   mv "$in_progress_file" "$CFE_FR_TRANSPORT_DIR/$CFE_FR_FEEDER.sql.$CFE_FR_COMPRESSOR_EXT"
   chown "$CFE_FR_FEEDER_USERNAME" "$CFE_FR_TRANSPORT_DIR/$CFE_FR_FEEDER.sql.$CFE_FR_COMPRESSOR_EXT"
 fi
+
+if [ -n "CFE_FR_SUPERHUB_HOSTKEYS" ]; then
```
Comment (craigcomstock) [inline templates/federated_reporting/dump.sh:80]: Not sure if I really need this if here. The dump.sh file should be linked to the policy which provides the data.

Comment (craigcomstock) [reply templates/federated_reporting/dump.sh:80]: e.g. even with only one superhub for this feeder, `$CFE_FR_SUPERHUB_HOSTKEYS` would be defined and so the message might change to

```
log "Copying for superhub(s): $CFE_FR_SUPERHUB_HOSTKEYS"
```

File: cfe_internal/enterprise/federation/federation.cf [hunk 2]
```diff
@@ -96,6 +96,13 @@ bundle agent config
                        strcmp( "push_over_rsync", "$(data[remote_hubs][$(remotes)][transport][mode])"));
 
     am_feeder::
+      # List of superhub hostkeys for use with dump process
+      "superhubs[$(remotes)]" string => "$(data[remote_hubs][$(remotes)][hostkey])",
+          if => and( strcmp( "on", "$(data[remote_hubs][$(remotes)][target_state])" ),
+                     strcmp( "superhub", "$(data[remote_hubs][$(remotes)][role])" ),
+                     strcmp( "pull_over_rsync", "$(data[remote_hubs][$(remotes)][transport][mode])"));
+      "superhub_hostkeys" string => join(" ", getvalues(superhubs));
+
```
Comment (nickanderson) [inline cfe_internal/enterprise/federation/federation.cf:105]: Makes sense.

File: cfe_internal/enterprise/federation/federation.cf [hunk 3]
```diff
@@ -442,6 +449,10 @@ bundle agent federation_manage_files
       string => "$(cfengine_enterprise_federation:config.transport_user)";
     "login" data => parsejson('{"login":"$(cfengine_enterprise_federation:config.login)"}');
     "feeder_username" data => parsejson('{"feeder_username":"$(cfengine_enterprise_federation:config.transport_user)"}');
+    "superhub_hostkeys" string => ifelse( isvariable("cfengine_enterprise_federatin:config.superhub_hostkeys"),
+                                        "$(cfengine_enterprise_federation:config.superhub_hostkeys)",
+                                        "" );
+    "this_hostkey" data => parsejson('{"this_hostkey":"$(default:sys.key_digest)"}');
```
Comment (nickanderson) [inline cfe_internal/enterprise/federation/federation.cf:455]: Makes sense.

File: templates/federated_reporting/config.sh.mustache [hunk 4]
```diff
@@ -58,6 +58,10 @@ CFE_FR_SUPERHUB_IMPORT_DIR="{{import_dir}}"
 CFE_FR_SUPERHUB_IMPORT_DIR="${CFE_FR_SUPERHUB_IMPORT_DIR:-$DEFAULT_PREFIX/superhub/import}"
 CFE_FR_SUPERHUB_LOGIN="{{login}}"
 CFE_FR_SUPERHUB_LOGIN="${CFE_FR_SUPERHUB_LOGIN:-cftransport@superhub}"
+CFE_FR_SUPERHUB_HOSTKEYS="{{superhub_hostkeys}}"
+CFE_FR_SUPERHUB_HOSTKEYS="${CFE_FR_SUPERHUB_HOSTKEYS:-superhubhostkeys}"
+CFE_FR_THIS_HOSTKEY="{{this_hostkey}}"
+CFE_FR_THIS_HOSTKEY="${CFE_FR_THIS_HOSTKEY:-myhostkey}"
```
Comment (nickanderson) [inline templates/federated_reporting/config.sh.mustache:64]: OK.

File: templates/federated_reporting/dump.sh (hunk 1 above)
Comment (nickanderson) [reply templates/federated_reporting/dump.sh:80]: +1, I like the suggested change in log message.

File: templates/federated_reporting/dump.sh [hunk 5]
```diff
@@ -76,3 +76,13 @@ else
...
+  log "Copying for multiple superhubs: $CFE_FR_SUPERHUB_HOSTKEYS"
+  for superhub_hostkey in $CFE_FR_SUPERHUB_HOSTKEYS; do
+    mkdir -p "$CFE_FR_TRANSPORT_DIR/$superhub_hostkey"
+    cp "$CFE_FR_TRANSPORT_DIR/$CFE_FR_FEEDER.sql.$CFE_FR_COMPRESSOR_EXT" "$CFE_FR_TRANSPORT_DIR/$superhub_hostkey/"
+    chown -R "$CFE_FR_FEEDER_USERNAME" "$CFE_FR_TRANSPORT_DIR/$superhub_hostkey"
+  done
+  log "Copying for multiple superhubs: DONE"
```
Comment (nickanderson) [inline templates/federated_reporting/dump.sh:87]: If you change the log message, this one could be tweaked as well.

File: templates/federated_reporting/dump.sh [hunk 6]
```diff
@@ -76,3 +76,13 @@ else
   mv "$in_progress_file" "$CFE_FR_TRANSPORT_DIR/$CFE_FR_FEEDER.sql.$CFE_FR_COMPRESSOR_EXT"
   chown "$CFE_FR_FEEDER_USERNAME" "$CFE_FR_TRANSPORT_DIR/$CFE_FR_FEEDER.sql.$CFE_FR_COMPRESSOR_EXT"
 fi
+
+if [ -n "CFE_FR_SUPERHUB_HOSTKEYS" ]; then
+  log "Copying for multiple superhubs: $CFE_FR_SUPERHUB_HOSTKEYS"
+  for superhub_hostkey in $CFE_FR_SUPERHUB_HOSTKEYS; do
+    mkdir -p "$CFE_FR_TRANSPORT_DIR/$superhub_hostkey"
+    cp "$CFE_FR_TRANSPORT_DIR/$CFE_FR_FEEDER.sql.$CFE_FR_COMPRESSOR_EXT" "$CFE_FR_TRANSPORT_DIR/$superhub_hostkey/"
```
Comment (nickanderson) [inline templates/federated_reporting/dump.sh:84]: Do I understand correctly that this makes a copy of the data for each superhub?

So, say the dump is 10G, if you feed 2 superhubs, now you will need 20G of disk on the feeder?

If so, perhaps we make hard links?

**Review by nickanderson — APPROVED** [review]
Seems OK to me, especially for an initial stab.

One comment/question about how this current implementation changes storage requirements for each feeder ...

File: templates/federated_reporting/dump.sh (hunk 6 above)
Comment (craigcomstock) [reply templates/federated_reporting/dump.sh]: done. seems to work.

**Comment by craigcomstock** [comment]
Looks good. I will let burn-in on some machines I have this running on. So far so good. (two deb11 superhubs x86 laptops, one deb11 feeder rpi4 and a few other clients)

**Comment by craigcomstock** [comment]
cherry picked only to 3.18.x and not 3.15.x: https://github.com/cfengine/masterfiles/pull/2489

**Pull request merged**